lib/io.js

```diff
@@ -51,22 +51,31 @@ BinaryDecoder.prototype = {
         var oldOffset = this._input.offset;
         var b = this.readByte();
         if (b instanceof AvroErrors.BlockDelayReadError) return b;
+        b = Long.fromBits(b, 0);
 
-        var n = b & 0x7F;
+        var L_0x7f = Long.fromNumber(0x7F);
+        var L_0xFF = Long.fromNumber(0xFF);
+
+        var n = b.and(L_0x7f);
         var shift = 7;
+        var numShifts = 0;
 
-        while ((b & 0x80) != 0) {
+        while (b.greaterThan(L_0x7f) && numShifts < 7) {
             b = this.readByte();
             if (b instanceof AvroErrors.BlockDelayReadError) {
                 //console.error("offset now is %d, old was %d, shift is %d", this._input.offset, oldOffset, shift);
                 this._input.skip(oldOffset - this._input.offset);
                 return b;
             }
-            n |= (b & 0x7F) << shift
-            shift += 7
+
+            b = Long.fromBits(b, 0);
+            n = n.xor(b.and(L_0x7f).shiftLeft(shift));
+
+            shift += 7;
+            numShifts++;
         }
 
-        return (n >> 1) ^ -(n & 1);
+        return (n.shiftRightUnsigned(1)).xor(-(n.and(1))).toNumber();
```

Comment:
There are a lot of things that would be nice if JavaScript would do, but then we'd probably be using a different language...